java-checks/src/main/java/org/sonar/java/checks/SwitchLastCaseIsDefaultCheck.java

```diff
@@ -42,35 +46,57 @@
 
   @Override
   public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
     SwitchStatementTree switchStatementTree = (SwitchStatementTree) tree;
-    CaseLabelTree defaultLabel = getDefaultLabel(switchStatementTree);
-    CaseLabelTree lastLabel = getLastLabel(switchStatementTree);
-    if (defaultLabel == null) {
-      reportIssue(switchStatementTree.switchKeyword(), "Add a default case to this switch.");
-    } else if (!defaultLabel.equals(lastLabel)) {
-      reportIssue(defaultLabel, "Move this default to the end of the switch.");
+    Optional<CaseLabelTree> defaultLabel = getDefaultLabel(switchStatementTree);
+    if (!defaultLabel.isPresent()) {
+      if (!isSwitchOnEnum(switchStatementTree)) {
+        reportIssue(switchStatementTree.switchKeyword(), "Add a default case to this switch.");
+      } else if (missingCasesOfEnum(switchStatementTree)) {
+        reportIssue(switchStatementTree.switchKeyword(), "Complete cases by adding the missing enum constants or add a default case to this switch.");
+      }
+    } else if (!defaultLabel.get().equals(getLastLabel(switchStatementTree))) {
+      reportIssue(defaultLabel.get(), "Move this default to the end of the switch.");
     }
   }
 
+  private static Optional<CaseLabelTree> getDefaultLabel(SwitchStatementTree switchStatementTree) {
+    return allLabels(switchStatementTree).stream()
+      .filter(SwitchLastCaseIsDefaultCheck::isDefault)
+      .findAny();
+  }
+
+  private static boolean isDefault(CaseLabelTree caseLabelTree) {
+    return JavaKeyword.DEFAULT.getValue().equals(caseLabelTree.caseOrDefaultKeyword().text());
+  }
 
-  private static CaseLabelTree getDefaultLabel(SwitchStatementTree switchStatementTree) {
-    for (CaseGroupTree caseGroupTree : switchStatementTree.cases()) {
-      for (CaseLabelTree caseLabelTree : caseGroupTree.labels()) {
-        if (JavaKeyword.DEFAULT.getValue().equals(caseLabelTree.caseOrDefaultKeyword().text())) {
-          return caseLabelTree;
-        }
-      }
-    }
-    return null;
+  private static boolean isSwitchOnEnum(SwitchStatementTree switchStatementTree) {
+    return switchStatementTree.expression().symbolType().symbol().isEnum();
+  }
+
+  private static boolean missingCasesOfEnum(SwitchStatementTree switchStatementTree) {
+    return numberConstants(switchStatementTree) > allLabels(switchStatementTree).size();
+  }
+
+  private static List<CaseLabelTree> allLabels(SwitchStatementTree switchStatementTree) {
+    return switchStatementTree.cases().stream()
+      .flatMap(caseGroup -> caseGroup.labels().stream())
+      .collect(Collectors.toList());
```

Comment:
Given that you use it solely as a stream or to gets its size you could return a Stream here... no need to collect that as a list...